File: src/lib/stores/releases.ts
/**
 * Releases store — list fetching, selection with detail load, CRUD mutations,
 * asset upload tracking via TaskManager, and cache for xrefs cross-linking.
 *
 * Asset uploads are fire-and-forget: `doUploadAsset` returns a TaskId. The
 * caller should subscribe to tasks.ts events for progress; on task completion
 * the detail is refreshed automatically via `refreshSelectedDetail()`.
 *
 * Mirrors the pattern used by `issues.ts` and `mr-pr.ts`.
 */

import { getErrorMessage } from "$lib/api/errors";
import { writable, derived, get } from "svelte/store";
import type {
  Release,
  ReleaseDetail,
  TaskId,
  CreateReleaseInput,
  EditReleasePatch,
} from "../types";
import {
  listReleases as apiList,
  getReleaseDetail as apiDetail,
  createRelease as apiCreate,
  editRelease as apiEdit,
  deleteRelease as apiDelete,
  publishRelease as apiPublish,
  uploadReleaseAsset as apiUpload,
  deleteReleaseAsset as apiDeleteAsset,
  createTagAndRelease as apiCreateTagAndRelease,
} from "../api/tauri";
import { runMutation } from "../api/runMutation";
import { fetchListIntoStore } from "../utils/store-helpers";
import { withTimeout } from "../utils/withTimeout";
import { addToast } from "./toast";
import * as m from "$lib/paraglide/messages";

/**
 * Timeout for the release detail fetch. Mirrors the MR/PR detail
 * timeout so slow `gh api releases/tags/...` calls can't strand the
 * detail pane in a permanent loading state.
 */
const RELEASE_DETAIL_TIMEOUT_MS = 15_000;

/** Current list of releases for the active repository. */
export const releases = writable<Release[]>([]);
/** Whether the list is currently loading. */
export const releasesLoading = writable(false);
/** Currently selected release tag (null = nothing selected). */
export const selectedReleaseTag = writable<string | null>(null);
/** Full detail of the selected release. */
export const releaseDetail = writable<ReleaseDetail | null>(null);
/** Whether the detail view is loading. */
export const releaseDetailLoading = writable(false);
/**
 * Last error raised while loading the selected release detail. Null
 * on success or before any load is attempted. `ReleaseDetail.svelte`
 * reads this store via `ForgeDetailShell` to render an inline banner
 * with a Retry affordance.
 */
export const releaseDetailError = writable<string | null>(null);

/** Set of release tags (for `xrefs.ts` cross-linking of `vX.Y.Z` strings). */
export const releaseTagSet = derived(
  releases,
  ($releases) => new Set($releases.map((r) => r.tag)),
);

/** Active upload task IDs keyed by release tag (for UI progress rows). */
export const activeUploads = writable<Map<string, Set<TaskId>>>(new Map());

/** Fetch the releases list (newest 30). Replaces current list. */
export async function refreshReleases(): Promise<void> {
  await fetchListIntoStore(
    releases,
    releasesLoading,
    selectedReleaseTag,
    () => apiList(30),
    [],
    (r) => r.tag,
  );
  if (get(selectedReleaseTag) === null) {
    releaseDetail.set(null);
  }
}

/**
 * Select a release and load its detail.
 *
 * The fetch is raced against a {@link RELEASE_DETAIL_TIMEOUT_MS}
 * timer so a hung IPC call can't leave the detail pane stuck on a
 * spinner. On any failure (including timeout) the error is written
 * to `releaseDetailError` and surfaced via a toast. Every async
 * side-effect is gated on the selected tag still matching the one
 * this call was started for, so rapid-fire clicks don't race.
 */
export function selectRelease(tag: string): void {
  selectedReleaseTag.set(tag);
  releaseDetail.set(null);
  releaseDetailError.set(null);
  releaseDetailLoading.set(true);
  const expected = tag;
  withTimeout(apiDetail(tag), RELEASE_DETAIL_TIMEOUT_MS)
    .then((d) => {
      if (get(selectedReleaseTag) !== expected) return;
      releaseDetail.set(d);
    })
    .catch((err) => {
      if (get(selectedReleaseTag) !== expected) return;
      const msg = getErrorMessage(err);
      releaseDetail.set(null);
      releaseDetailError.set(msg);
      addToast({
        message: m.release_load_failed({ tag, error: msg }),
        type: "error",
      });
    })
    .finally(() => {
      if (get(selectedReleaseTag) !== expected) return;
      releaseDetailLoading.set(false);
    });
}

/**
 * Force-reload the currently selected detail (called after edits and
 * asset uploads finish). Uses the same 15 s timeout as
 * {@link selectRelease}, but does **not** fire a toast on failure —
 * this is a background refresh, not a user-initiated load, so the
 * error is only written to `releaseDetailError` for the banner to
 * pick up. The user can still manually retry.
 */
export async function refreshSelectedDetail(): Promise<void> {
  const tag = get(selectedReleaseTag);
  if (!tag) return;
  try {
    const d = await withTimeout(apiDetail(tag), RELEASE_DETAIL_TIMEOUT_MS);
    if (get(selectedReleaseTag) === tag) {
      releaseDetail.set(d);
      releaseDetailError.set(null);
    }
  } catch (err) {
    if (get(selectedReleaseTag) !== tag) return;
    const msg = getErrorMessage(err);
    releaseDetailError.set(msg);
    /* no toast — this is an auto-refresh path. */
  }
}

/** Create a release via the provider CLI and refresh the list. */
export async function doCreateRelease(
  input: CreateReleaseInput,
): Promise<Release> {
  const created = await runMutation({
    kind: "release_create",
    invoke: () => apiCreate(input),
    successToast: (r) => `Created release ${r.tag}`,
    failureToastPrefix: "Release create failed",
  });
  await refreshReleases();
  return created;
}

/** Atomic create-tag + push + create-release. Returns the TaskId. */
export async function doCreateTagAndRelease(
  tag: string,
  sourceRef: string,
  remote: string,
  input: CreateReleaseInput,
): Promise<TaskId> {
  // Long-running task — progress + completion are reported by the
  // Rust-side TaskManager, which fires its own task entries. That needed an
  // explicit `Background` kind to be true: as a `Generic` task it was dropped
  // by `should_emit` and no entry was ever fired.
  // Toast policy still runs through runMutation so a provider-CLI
  // failure (e.g. `gh` not installed) surfaces a sticky error.
  return runMutation({
    kind: "release_create_tag_and_release",
    invoke: () => apiCreateTagAndRelease(tag, sourceRef, remote, input),
    failureToastPrefix: "Release create failed",
  });
}

/** Edit a release, then refresh list + detail. */
export async function doEditRelease(
  tag: string,
  patch: EditReleasePatch,
): Promise<void> {
  await runMutation({
    kind: "release_edit",
    invoke: () => apiEdit(tag, patch),
    successToast: () => `Updated release ${tag}`,
    failureToastPrefix: "Release edit failed",
  });
  await refreshReleases();
  await refreshSelectedDetail();
}

/** Delete a release and refresh. Clears selection if it was the deleted tag. */
export async function doDeleteRelease(tag: string): Promise<void> {
  await runMutation({
    kind: "release_delete",
    invoke: () => apiDelete(tag),
    successToast: () => `Deleted release ${tag}`,
    failureToastPrefix: "Release delete failed",
  });
  if (get(selectedReleaseTag) === tag) {
    selectedReleaseTag.set(null);
    releaseDetail.set(null);
  }
  await refreshReleases();
}

/** Publish a draft release (GitHub only), then refresh. */
export async function doPublishRelease(tag: string): Promise<void> {
  await runMutation({
    kind: "release_publish",
    invoke: () => apiPublish(tag),
    successToast: () => `Published ${tag}`,
    failureToastPrefix: "Release publish failed",
    trackAsTask: true,
  });
  await refreshReleases();
  await refreshSelectedDetail();
}

/**
 * Upload an asset to a release. Returns the TaskId immediately; the caller
 * should show progress via `<AssetUploadProgress>` and call `completeUpload`
 * when the task finishes to clean up the active-uploads map.
 */
export async function doUploadAsset(
  tag: string,
  assetPath: string,
  label?: string,
): Promise<TaskId> {
  const id = await apiUpload(tag, assetPath, label);
  activeUploads.update((map) => {
    const next = new Map(map);
    const set = new Set(next.get(tag) ?? []);
    set.add(id);
    next.set(tag, set);
    return next;
  });
  return id;
}

/** Remove an upload task from the active-uploads map. */
export function completeUpload(tag: string, taskId: TaskId): void {
  activeUploads.update((map) => {
    const next = new Map(map);
    const set = new Set(next.get(tag) ?? []);
    set.delete(taskId);
    if (set.size === 0) {
      next.delete(tag);
    } else {
      next.set(tag, set);
    }
    return next;
  });
}

/** Delete a release asset and refresh the detail view. */
export async function doDeleteAsset(
  tag: string,
  assetId: number,
): Promise<void> {
  await runMutation({
    kind: "release_asset_delete",
    invoke: () => apiDeleteAsset(tag, assetId),
    successToast: () => "Asset deleted",
    failureToastPrefix: "Asset delete failed",
  });
  await refreshSelectedDetail();
}

/** Reset all release state (on project switch). */
export function clearReleaseState(): void {
  releases.set([]);
  selectedReleaseTag.set(null);
  releaseDetail.set(null);
  releaseDetailLoading.set(false);
  releaseDetailError.set(null);
  activeUploads.set(new Map());
}

File: crates/app-core/src/commands/releases.rs
//! Release listing, detail, CRUD, and asset operations.
//!
//! Read + write operations (list/get/create/edit/delete/publish,
//! delete_asset) are synchronous-style commands that dispatch to the active
//! [`forge_provider::ForgeProvider`] on a `spawn_blocking` thread.
//!
//! **Asset upload is special:** it returns a `TaskId` immediately and
//! streams stdout/stderr from the `gh release upload` / `glab release
//! upload` subprocess via the [`task_runner::TaskManager`]. Large binary
//! uploads don't block the UI; the frontend subscribes to task events to
//! show progress.
//!
//! ## Atomic create-tag + push + release
//!
//! [`create_tag_and_release`] creates a local tag, pushes it to the
//! remote, then creates the release — all sequentially inside a single
//! task so any failure surfaces with a single error. It is used by the
//! "new tag" mode of the `CreateReleaseDialog` so the user gets atomic
//! feedback on the full flow.

use std::sync::Arc;

use forge_provider::{
    CreateReleaseInput, EditReleasePatch, ForgeProvider, Release, ReleaseAsset, ReleaseDetail,
};
use mutation_events::MutationKind;
use task_runner::{SpawnOptions, TaskId, TaskKind, TaskManager};
use tauri::{AppHandle, State};

use super::helpers::*;
use crate::ipc_error::IpcError;
use crate::state::AppState;

/// List releases for the current repository, newest first.
#[tauri::command]
pub async fn list_releases(
    limit: Option<u32>,
    state: State<'_, AppState>,
) -> Result<Vec<Release>, IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    let limit = limit.unwrap_or(30);
    run_blocking(move || {
        provider
            .list_releases(limit)
            .map_err(|e| IpcError::from(e.to_string()))
    })
    .await
}

/// Fetch full detail (body + assets) for a single release by tag.
#[tauri::command]
pub async fn get_release_detail(
    tag: String,
    state: State<'_, AppState>,
) -> Result<ReleaseDetail, IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .get_release(&tag)
            .map_err(|e| IpcError::from(e.to_string()))
    })
    .await
}

/// List just the asset records for a release.
#[tauri::command]
pub async fn list_release_assets(
    tag: String,
    state: State<'_, AppState>,
) -> Result<Vec<ReleaseAsset>, IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .list_release_assets(&tag)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(IpcError::from)
}

/// Create a new release.
///
/// On GitHub, `input.target_commit` can be an unpushed branch or SHA — the
/// CLI will create the tag remotely. On GitLab the caller is expected to
/// have pushed the tag already (use [`create_tag_and_release`] for the
/// create+push+release flow).
#[tauri::command]
pub async fn create_release(
    input: CreateReleaseInput,
    state: State<'_, AppState>,
) -> Result<Release, IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .create_release(input)
            .map_err(|e| IpcError::from(e.to_string()))
    })
    .await
}

/// Edit a release's title, body, and/or draft/prerelease flags.
#[tauri::command]
pub async fn edit_release(
    tag: String,
    patch: EditReleasePatch,
    state: State<'_, AppState>,
) -> Result<(), IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .edit_release(&tag, patch)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(IpcError::from)
}

/// Delete a release. The underlying tag is not removed.
#[tauri::command]
pub async fn delete_release(tag: String, state: State<'_, AppState>) -> Result<(), IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .delete_release(&tag)
            .map_err(|e| IpcError::from(e.to_string()))
    })
    .await
}

/// Publish a draft release. GitHub only — GitLab returns a NotSupported error.
///
/// Wraps the provider call inside a
/// [`MutationGuard`][mutation_events::MutationGuard] scope so that on success a
/// `project-mutated` event with [`MutationKind::TagCreate`] is emitted —
/// publishing a release makes its tag ref visible on the remote.
#[tauri::command]
pub async fn publish_release(
    tag: String,
    state: State<'_, AppState>,
    app: AppHandle,
) -> Result<(), IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    with_mutation_guard_async(&state, &app, MutationKind::TagCreate, || async move {
        run_blocking(move || {
            provider
                .publish_release(&tag)
                .map_err(|e| IpcError::from(e.to_string()))
        })
        .await
    })
    .await
}

/// Delete a single release asset by ID.
#[tauri::command]
pub async fn delete_release_asset(
    tag: String,
    asset_id: u64,
    state: State<'_, AppState>,
) -> Result<(), IpcError> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .delete_release_asset(&tag, asset_id)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(IpcError::from)
}

/// Upload a binary asset to a release.
///
/// Non-blocking: returns a `TaskId` immediately. The underlying `gh
/// release upload` / `glab release upload` subprocess runs via
/// [`TaskManager`] so its stdout/stderr stream to the task popover and
/// the UI stays responsive even for large binaries.
///
/// The frontend subscribes to `task-completed` / `task-failed` events for
/// this id and re-fetches the release detail on success to pick up the
/// newly uploaded asset row.
#[tauri::command]
pub async fn upload_release_asset(
    tag: String,
    asset_path: String,
    label: Option<String>,
    state: State<'_, AppState>,
    task_manager: State<'_, Arc<TaskManager>>,
) -> Result<TaskId, IpcError> {
    let cwd = get_active_project_path(&state)?;

    // Resolve which CLI we're speaking to and the upload argv shape.
    let (binary, args) = {
        let kind = {
            let providers = state.providers.lock().map_err(|e| e.to_string())?;
            let active = state
                .active_provider_index
                .lock()
                .map_err(|e| e.to_string())?;
            let idx = active.ok_or_else(|| "No active provider".to_string())?;
            let conn = providers
                .get(idx)
                .ok_or_else(|| "Active provider index out of bounds".to_string())?;
            conn.kind
        };
        let bin = resolve_cli_binary(&state, kind)?;
        // Treat empty-string labels as "no label" so we never produce a
        // dangling `path#` in the gh argv.
        let label_ref = label.as_deref().filter(|s| !s.is_empty());
        let args = match kind {
            provider::ProviderKind::GitHub => {
                cli_provider::build_gh_upload_args(&tag, &asset_path, label_ref)
            }
            provider::ProviderKind::GitLab => {
                cli_provider::build_glab_upload_args(&tag, &asset_path, label_ref)
            }
        };
        (bin, args)
    };

    let file_name = std::path::Path::new(&asset_path)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or("asset")
        .to_string();
    let task_label = match label.as_deref() {
        Some(l) if !l.is_empty() => format!("Upload asset: {file_name} ({l}) → {tag}"),
        _ => format!("Upload asset: {file_name} → {tag}"),
    };

    let args_ref: Vec<&str> = args.iter().map(String::as_str).collect();
    let binary_str = binary.to_string_lossy().to_string();

    // Explicit `Background`: the bare `spawn` tags `Generic`, which
    // `should_emit` drops, so the upload never reached the drawer. It has its
    // own `AssetUploadProgress` surface, but the drawer is where the user
    // looks for "what is running", and a large upload belongs there too.
    let id = task_manager
        .spawn_with_options(SpawnOptions {
            label: task_label,
            command: &binary_str,
            args: &args_ref,
            cwd: &cwd,
            cancellable: true,
            kind: TaskKind::Background,
            stdin: None,
        })
        .await;
    Ok(id)
}

/// Atomic create-tag + push + create-release.
///
/// Runs sequentially:
/// 1. Create a local tag pointing at `source_ref` (lightweight tag).
/// 2. Push the tag to `remote` via `git push`.
/// 3. Call the provider's `create_release`.
///
/// Returns immediately with a `TaskId`; progress messages are streamed as
/// task output via [`TaskManager::spawn`]. Because the underlying steps use
/// heterogeneous mechanisms (git CLI for steps 1-2, forge provider for step
/// 3), we assemble a small shell wrapper: the tag create/push runs via
/// `git` in a subprocess; the release create is then attempted in-process
/// after the subprocess exits. For simplicity and to fit into the
/// subprocess-only TaskManager, we shell out all three steps — tag +
/// push use `git`, and the release create is handled in a follow-up
/// blocking call after the task completes.
///
/// For now we take the simpler approach: run `git tag && git push` as a
/// single combined task, and perform the `create_release` call in the
/// success path via a polling listener, then emit a `release-created`
/// event. This mirrors the `checkout_mr_pr_locally` pattern.
#[tauri::command]
pub async fn create_tag_and_release(
    tag: String,
    source_ref: String,
    remote: String,
    input: CreateReleaseInput,
    state: State<'_, AppState>,
    task_manager: State<'_, Arc<TaskManager>>,
    app_handle: tauri::AppHandle,
) -> Result<TaskId, IpcError> {
    let cwd = get_active_project_path(&state)?;
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;

    // Step 1+2 as a single combined subprocess: `sh -c "git tag X SRC && git push REMOTE X"`.
    // This keeps stdout/stderr streaming through TaskManager.
    let tag_ref = format!("refs/tags/{tag}");
    let combined = format!(
        "git tag {tag} {source_ref} && git push {remote} {tag_ref}",
        tag = shell_escape(&tag),
        source_ref = shell_escape(&source_ref),
        remote = shell_escape(&remote),
        tag_ref = shell_escape(&tag_ref),
    );

    let (shell, flag) = if cfg!(target_os = "windows") {
        ("cmd", "/C")
    } else {
        ("sh", "-c")
    };
    let args = [flag, combined.as_str()];

    // Explicit `Background`: with the bare `spawn`'s `Generic` kind, the
    // doc-comment on `doCreateTagAndRelease` ("progress + completion are
    // reported by the Rust-side TaskManager, which already fires its own task
    // entries") was not true — `should_emit` dropped every one of them.
    let id = task_manager
        .spawn_with_options(SpawnOptions {
            label: format!("Create tag + release: {tag}"),
            command: shell,
            args: &args,
            cwd: &cwd,
            cancellable: true,
            kind: TaskKind::Background,
            stdin: None,
        })
        .await;

    // Spawn a listener that, on success, invokes `create_release` and emits
    // a `release-created` event with the Release as payload.
    let tm: Arc<TaskManager> = Arc::clone(&task_manager);
    let handle = app_handle.clone();
    let tag_for_listener = tag.clone();
    tokio::spawn(async move {
        use task_runner::TaskStatus;
        // Failed / Cancelled / NotFound: nothing to emit (the task log already
        // captures the error). Silent exit matches the prior behaviour.
        if let Ok(TaskStatus::Completed) = tm.wait_for_terminal(id).await {
            let provider = Arc::clone(&provider);
            let input = input.clone();
            let tag = tag_for_listener.clone();
            let result = tokio::task::spawn_blocking(move || {
                provider.create_release(input).map_err(|e| e.to_string())
            })
            .await;
            use tauri::Emitter as _;
            match result {
                Ok(Ok(release)) => {
                    let _ = handle.emit("release-created", &release);
                }
                Ok(Err(e)) => {
                    let _ = handle.emit(
                        "release-create-failed",
                        &serde_json::json!({ "tag": tag, "error": e }),
                    );
                }
                Err(e) => {
                    let _ = handle.emit(
                        "release-create-failed",
                        &serde_json::json!({ "tag": tag, "error": e.to_string() }),
                    );
                }
            }
        }
    });

    Ok(id)
}

#[cfg(test)]
mod tests {
    //! Tests the release trait surface through `MockProvider` (defaults to
    //! NotSupported) and the `cli_provider` argv-builders that the
    //! `upload_release_asset` command composes. The async task-runner glue
    //! is covered at the `task_runner` crate level.

    use forge_provider::mock::MockProvider;
    use forge_provider::{
        CreateReleaseInput, EditReleasePatch, ForgeError, ForgeKind, ForgeProvider,
    };

    #[test]
    fn mock_provider_list_releases_returns_not_supported() {
        let provider = MockProvider::new(ForgeKind::GitHub);
        assert!(matches!(
            provider.list_releases(30),
            Err(ForgeError::NotSupported)
        ));
    }

    #[test]
    fn mock_provider_create_release_returns_not_supported() {
        let provider = MockProvider::new(ForgeKind::GitLab);
        let input = CreateReleaseInput {
            tag: "v1.0.0".into(),
            target_commit: "".into(),
            name: "1.0".into(),
            body: "Notes".into(),
            draft: false,
            prerelease: false,
            generate_notes: false,
        };
        assert!(matches!(
            provider.create_release(input),
            Err(ForgeError::NotSupported)
        ));
    }

    #[test]
    fn mock_provider_edit_release_returns_not_supported() {
        let provider = MockProvider::new(ForgeKind::GitHub);
        let patch = EditReleasePatch {
            name: Some("renamed".into()),
            body: None,
            draft: Some(false),
            prerelease: None,
        };
        assert!(matches!(
            provider.edit_release("v1.0.0", patch),
            Err(ForgeError::NotSupported)
        ));
    }

    #[test]
    fn mock_provider_delete_release_returns_not_supported() {
        let provider = MockProvider::new(ForgeKind::GitHub);
        assert!(matches!(
            provider.delete_release("v1.0.0"),
            Err(ForgeError::NotSupported)
        ));
    }

    #[test]
    fn build_gh_upload_args_includes_tag_and_path() {
        let args = cli_provider::build_gh_upload_args("v1.0.0", "/tmp/a.zip", None);
        // The argv must include the tag and the path somewhere — label is
        // None so the path should appear without a "#label" suffix.
        assert!(
            args.iter().any(|a| a == "v1.0.0"),
            "gh upload argv must include the tag, got {args:?}"
        );
        assert!(
            args.iter().any(|a| a.ends_with("a.zip")),
            "gh upload argv must reference the asset path, got {args:?}"
        );
    }

    #[test]
    fn build_glab_upload_args_ignores_label_but_includes_tag_and_file() {
        // glab's `release upload` subcommand does not accept a label
        // argument (unlike `gh release upload`), so the label is dropped
        // by design. Verify both the tag and file path make it through.
        let args = cli_provider::build_glab_upload_args("v2", "/tmp/b.bin", Some("Linux binary"));
        assert!(
            args.iter().any(|a| a == "v2"),
            "glab upload argv must include the tag, got {args:?}"
        );
        assert!(
            args.iter().any(|a| a.ends_with("b.bin")),
            "glab upload argv must reference the asset path, got {args:?}"
        );
    }
}
